lib/state/cachingaccesspoint.go

```diff
@@ -371,110 +440,161 @@ func (cs *CachingAuthClient) GetReverseTunnels() (tunnels []services.ReverseTunn
+			return err
+		},
+		updateCache: func() (keys []string, cerr error) {
+			if err := cs.trust.DeleteCertAuthority(id); err != nil {
+				if !trace.IsNotFound(err) {
+					return nil, trace.Wrap(err)
+				}
+			}
+			cs.setTTL(ca)
+			if err := cs.trust.UpsertCertAuthority(ca); err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return
+		},
 	})
-	if err != nil {
-		if trace.IsConnectionProblem(err) {
-			return cs.trust.GetCertAuthority(id, loadKeys)
-		}
-		return ca, err
-	}
-	if err := cs.trust.DeleteCertAuthority(id); err != nil {
-		if !trace.IsNotFound(err) {
-			return nil, trace.Wrap(err)
-		}
-	}
-	cs.setTTL(ca)
-	if err := cs.trust.UpsertCertAuthority(ca); err != nil {
-		return nil, trace.Wrap(err)
+	return
+}
+
+func boolToString(b bool) string {
+	if b {
+		return "true"
 	}
-	return ca, nil
+	return "false"
+}
+
+func certsKey(ct services.CertAuthType, loadKeys bool) string {
+	return strings.Join([]string{"cas", string(ct), boolToString(loadKeys)}, "/")
+}
+
+func certKey(id services.CertAuthID, loadKeys bool) string {
+	return strings.Join([]string{"cas", string(id.Type), boolToString(loadKeys), id.DomainName}, "/")
 }
 
 // GetCertAuthorities is a part of auth.AccessPoint implementation
 func (cs *CachingAuthClient) GetCertAuthorities(ct services.CertAuthType, loadKeys bool) (cas []services.CertAuthority, err error) {
-	err = cs.try(func() error {
-		cas, err = cs.ap.GetCertAuthorities(ct, loadKeys)
-		return err
+	cs.fetch(params{
+		key: certsKey(ct, loadKeys),
+		fetch: func() error {
+			cas, err = cs.ap.GetCertAuthorities(ct, loadKeys)
+			return err
+		},
+		useCache: func() error {
+			cas, err = cs.trust.GetCertAuthorities(ct, loadKeys)
+			return err
+		},
+		updateCache: func() (keys []string, cerr error) {
+			if err := cs.trust.DeleteAllCertAuthorities(ct); err != nil {
+				if !trace.IsNotFound(err) {
+					return nil, trace.Wrap(err)
+				}
+			}
+			for _, ca := range cas {
+				cs.setTTL(ca)
+				if err := cs.trust.UpsertCertAuthority(ca); err != nil {
+					return nil, trace.Wrap(err)
+				}
+				keys = append(keys, certKey(ca.GetID(), loadKeys))
+			}
+			return
+		},
 	})
-	if err != nil {
-		if trace.IsConnectionProblem(err) {
-			return cs.trust.GetCertAuthorities(ct, loadKeys)
-		}
-		return cas, err
-	}
-	if err := cs.trust.DeleteAllCertAuthorities(ct); err != nil {
-		if !trace.IsNotFound(err) {
-			return nil, trace.Wrap(err)
-		}
-	}
-	for _, ca := range cas {
-		cs.setTTL(ca)
-		if err := cs.trust.UpsertCertAuthority(ca); err != nil {
-			return nil, trace.Wrap(err)
-		}
-	}
-	return cas, err
+	return
+}
+
+func usersKey() string {
+	return "users"
+}
+
+func userKey(username string) string {
+	return strings.Join([]string{"users", username}, "/")
 }
 
 // GetUsers is a part of auth.AccessPoint implementation
 func (cs *CachingAuthClient) GetUsers() (users []services.User, err error) {
-	err = cs.try(func() error {
-		users, err = cs.ap.GetUsers()
-		return err
+	cs.fetch(params{
```

Comment:
Errors from `cs.fetch` are ignored everywhere, maybe they should at least be logged inside?